### PR TITLE
Air

### PR DESCRIPTION
**Pull request by liruipeng**
pulled the new master in, and resubmit a PR

### CODE REVIEW

**Review by rfalgout — APPROVED** [review]
This looks good to me on the surface.  It might be good for you to give us a high-level overview of the changes.  If others agree, maybe we can schedule a brief meeting for you to do that?

**Comment by liruipeng** [comment]
> This looks good to me on the surface. It might be good for you to give us a high-level overview of the changes. If others agree, maybe we can schedule a brief meeting for you to do that?

Yes, I am fine with it. Thanks!

**Comment by liruipeng** [comment]
> This looks good to me on the surface. It might be good for you to give us a high-level overview of the changes. If others agree, maybe we can schedule a brief meeting for you to do that?

@rfalgout maybe we all can look at it in the LS meeting this Thursday?

**Comment by rfalgout** [comment]
> > This looks good to me on the surface. It might be good for you to give us a high-level overview of the changes. If others agree, maybe we can schedule a brief meeting for you to do that?
> 
> @rfalgout maybe we all can look at it in the LS meeting this Thursday?

I am okay with that, but maybe we should instead try to meet sooner so that the pull request doesn't drag on for too long.  Only the reviewers need to be involved.  Your call.

File: src/parcsr_ls/F90_HYPRE_parcsr_amg.c [hunk 1]
```diff
@@ -113,13 +113,13 @@ hypre_F90_IFACE(hypre_boomeramgsolvet, HYPRE_BOOMERAMGSOLVET)
 void
 hypre_F90_IFACE(hypre_boomeramgsetrestriction, HYPRE_BOOMERAMGSETRESTRICTION)
    ( hypre_F90_Obj *solver,
-     hypre_F90_Int *restr_par,
+     hypre_F90_Real *restr_par,
```
Comment (liruipeng) [inline src/parcsr_ls/F90_HYPRE_parcsr_amg.c]: not a real number

File: src/parcsr_ls/HYPRE_parcsr_amg.c [hunk 2]
```diff
@@ -677,9 +758,6 @@ HYPRE_BoomerAMGSetRelaxOrder( HYPRE_Solver  solver,
 
 /*--------------------------------------------------------------------------
  * HYPRE_BoomerAMGSetGridRelaxPoints
- * DEPRECATED.  There are memory management problems associated with the
```
Comment (liruipeng) [inline src/parcsr_ls/HYPRE_parcsr_amg.c]: put the comments back ....

File: src/parcsr_ls/HYPRE_parcsr_ls.h [hunk 3]
```diff
@@ -743,21 +758,21 @@ HYPRE_Int HYPRE_BoomerAMGSetCycleRelaxType(HYPRE_Solver  solver,
...
+ * The default is 0.
  **/
 HYPRE_Int HYPRE_BoomerAMGSetRelaxOrder(HYPRE_Solver  solver,
                                        HYPRE_Int     relax_order);
 
 /*
- * (Optional) Defines in which order the points are relaxed. 
+ * (Optional) Defines in which order the points are relaxed.
  *
- * Note: This routine will be phased out!!!!
```
Comment (liruipeng) [inline src/parcsr_ls/HYPRE_parcsr_ls.h]: put the note back in

File: src/parcsr_ls/par_amg_setup.c [hunk 4]
```diff
@@ -834,11 +837,12 @@ hypre_BoomerAMGSetup( void               *amg_vdata,
       if (level >= nodal_levels)
       {
          nodal = 0;
+         num_functions = 1;
```
Comment (liruipeng) [inline src/parcsr_ls/par_amg_setup.c]: why set num_functions = 1?

File: src/parcsr_ls/par_amg_setup.c [hunk 5]
```diff
@@ -1032,69 +1052,80 @@ hypre_BoomerAMGSetup( void               *amg_vdata,
...
+                                      CR_strong_th, 1,
+                                      num_functions, dof_func_array[level],&SCR);
+               hypre_BoomerAMGCoarsenCR(A_array[level], &CF_marker,
+                                        &coarse_size,
+                                        num_CR_relax_steps, IS_type, 1, grid_relax_type[0],
+                                        relax_weight[level], omega[level], CR_rate,
+                                        NULL,NULL,CR_use_CG,SCR);
+               hypre_ParCSRMatrixDestroy(SCR);
+            }
+            else if (coarsen_type == 999)
```
Comment (liruipeng) [inline src/parcsr_ls/par_amg_setup.c:1068]: remove debug code

File: src/parcsr_ls/par_amg_setup.c [hunk 6]
```diff
@@ -1725,38 +1776,38 @@ hypre_BoomerAMGSetup( void               *amg_vdata,
...
+            {
+               hypre_BoomerAMGBuildStdInterp(A_array[level], CF_marker,
+                                             S, coarse_pnts_global, num_functions, dof_func_array[level],
+                                             debug_flag, trunc_factor, P_max_elmts, sep_weight, col_offd_S_to_A, &P);
+               hypre_TFree(col_offd_S_to_A, HYPRE_MEMORY_HOST);
+            }
+            else if (interp_type == 100) /* 1pt interpolation */
+            {
+               hypre_BoomerAMGBuildInterpOnePnt(A_array[level], CF_marker, S,
+                                                coarse_pnts_global, 1, NULL,
```
Comment (liruipeng) [inline src/parcsr_ls/par_amg_setup.c:1790]: check num_functions

File: src/parcsr_ls/par_restr.c [hunk 7]
```diff
@@ -730,6 +818,189 @@ hypre_BoomerAMGBuildRestrAIR( hypre_ParCSRMatrix   *A,
...
+      for (j = 0; j < n; j++)
+      {
+         y[i] += x[j] * A[row0 + j];
+      }
+   }
+}
+
+// TODO : need to initialize and de-initialize GMRES
+
+static void fgmresT(HYPRE_Int n,
```
Comment (liruipeng) [inline src/parcsr_ls/par_restr.c]: duplicated fgmresT

File: src/parcsr_ls/par_strength.c [hunk 8]
```diff
@@ -1152,8 +1152,8 @@ hypre_BoomerAMGCreateSabs(hypre_ParCSRMatrix    *A,
       }
 
       /* compute row entries of S */
-      S_diag_j[A_diag_i[i]] = -1;
-      if ((fabs(row_sum) > fabs(diag)*max_row_sum) && (max_row_sum < 1.0))
+      S_diag_j[A_diag_i[i]] = -1; /* reject diag entry */
+      if ( fabs(row_sum) < fabs(diag)*(2.0-max_row_sum) && max_row_sum < 1.0 )
```
Comment (liruipeng) [inline src/parcsr_ls/par_strength.c:1156]: check history of this line

**Pull request merged**